internal/mssql/mock.go

```diff
@@ -0,0 +1,25 @@
+package mssql
+
+import (
+	"github.com/microsoft/go-sqlcmd/cmd/modern/sqlconfig"
+	"github.com/microsoft/go-sqlcmd/pkg/sqlcmd"
+)
+
+type MssqlType struct{}
+type MssqlMock struct{}
+
+func (m *MssqlMock) Connect(endpoint sqlconfig.Endpoint, user *sqlconfig.User, console sqlcmd.Console) *sqlcmd.Sqlcmd {
+	return &sqlcmd.Sqlcmd{
+		Exitcode:          0,
+		Connect:           nil,
+		Format:            nil,
+		Query:             "",
+		Cmd:               nil,
+		PrintError:        nil,
+		UnicodeOutputFile: false,
+	}
+}
+
+func (m *MssqlMock) Query(s *sqlcmd.Sqlcmd, text string) {
```

Comment:
This dependency should be defined by an interface instead of 2 types being defined in this package.

Whatever public methods you need for both MssqlMock and MssqlType should be defined by an interface, and the product code will implement the interface using MssqlType and the test code will implement it using whatever type it wants.

mssql.New would take an optional implementation of that interface and use MssqlType if not provided.

If the interface is public the implementation types can probably stay private.

Reply:
The PR was in draft in iteration 1, I was mid way though implementing... I have a mock impl available in non _test, so the unit tests can call `a sqlcmd install mssql` amd use dummy containers that don't even have SQL Server in them (the Hello World docker container), and then use this mock imple to fake out the Connect/SQL (that does the sa disable/create user account etc.), this keeps the unit testing runtime down to just a few seconds for all tests.